AsymmetricNetworkPartitionIT.shouldWithstandAsymmetricNetworkPartition

**Summary**

- How often does the test fail? at least twice
- Does it block your work? no
- Do we suspect that it is a real failure? I don't think so

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
java.lang.AssertionError: 

Expecting actual:
  0
not to be equal to:
  0

	at io.camunda.zeebe.it.clustering.network.AsymmetricNetworkPartitionIT.shouldWithstandAsymmetricNetworkPartition(AsymmetricNetworkPartitionIT.java:77)
</pre>
</details>

**Hypotheses**

None so far

**Logs**

None available. Failing build is: https://github.com/camunda/zeebe/actions/runs/2256830107

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
@@ -8,7 +8,6 @@
 package io.camunda.zeebe.it.clustering.network;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.fail;
 
 import com.github.dockerjava.api.command.CreateContainerCmd;
 import com.github.dockerjava.api.model.Capability;
@@ -20,8 +19,10 @@ import io.camunda.zeebe.client.api.response.Topology;
 import io.camunda.zeebe.test.util.testcontainers.ContainerLogsDumper;
 import io.camunda.zeebe.test.util.testcontainers.ZeebeTestContainerDefaults;
 import io.zeebe.containers.ZeebeBrokerNode;
+import io.zeebe.containers.ZeebeNode;
 import io.zeebe.containers.cluster.ZeebeCluster;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.net.URI;
 import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
@@ -34,7 +35,9 @@ import java.util.stream.Stream;
 import org.agrona.CloseHelper;
 import org.agrona.LangUtil;
 import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Named;
@@ -44,8 +47,6 @@ import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.testcontainers.containers.Container.ExecResult;
-import org.testcontainers.containers.GenericContainer;
 import org.testcontainers.containers.Network;
 import org.testcontainers.junit.jupiter.Container;
 import org.testcontainers.junit.jupiter.Testcontainers;
@@ -54,23 +55,23 @@ import org.testcontainers.junit.jupiter.Testcontainers;
 final class AsymmetricNetworkPartitionIT {
   private static final Logger LOGGER = LoggerFactory.getLogger(AsymmetricNetworkPartitionIT.class);
 
-  private final Network network = Network.newNetwork();
+  private static final Network NETWORK = Network.newNetwork();
 
   @Container
-  private final ZeebeCluster cluster =
+  private static final ZeebeCluster CLUSTER =
       ZeebeCluster.builder()
           .withImage(ZeebeTestContainerDefaults.defaultTestImage())
           .withBrokersCount(3)
           .withEmbeddedGateway(true)
           .withPartitionsCount(2)
           .withReplicationFactor(3)
-          .withBrokerConfig(this::configureBroker)
-          .withNetwork(network)
+          .withBrokerConfig(AsymmetricNetworkPartitionIT::configureBroker)
+          .withNetwork(NETWORK)
           .build();
 
   @SuppressWarnings("unused")
   @RegisterExtension
-  final ContainerLogsDumper logsWatcher = new ContainerLogsDumper(cluster::getBrokers, LOGGER);
+  final ContainerLogsDumper logsWatcher = new ContainerLogsDumper(CLUSTER::getBrokers, LOGGER);
 
   private ZeebeClient client;
 
@@ -81,14 +82,25 @@ final class AsymmetricNetworkPartitionIT {
         Arguments.arguments(Named.named("Message correlation", new MessageCorrelationTestCase())));
   }
 
+  @BeforeAll
+  static void beforeAll() {
+    CLUSTER.getBrokers().forEach((id, broker) -> installNetworkUtilities(broker));
+  }
+
+  @AfterAll
+  static void afterAll() {
+    CloseHelper.quietClose(NETWORK);
+  }
+
   @BeforeEach
   void beforeEach() {
-    client = cluster.newClientBuilder().build();
+    client = CLUSTER.newClientBuilder().build();
+    CLUSTER.getBrokers().forEach((id, broker) -> clearUnreachableRoutes(broker));
   }
 
   @AfterEach
   void afterEach() {
-    CloseHelper.quietCloseAll(client, network);
+    CloseHelper.quietCloseAll(client);
   }
 
   @DisplayName("Withstand Asymmetric Network Partition")
@@ -108,13 +120,13 @@ final class AsymmetricNetworkPartitionIT {
     final var firstLeader = getPartitionLeader(topology, 1);
     final var secondLeader = getPartitionLeader(topology, 2);
     final var firstLeaderIP =
-        getContainerNetworkIP(cluster.getBrokers().get(firstLeader.getNodeId()));
+        getContainerNetworkIP(CLUSTER.getBrokers().get(firstLeader.getNodeId()));
     testCase.given(client);
-    setupNetworkPartition(firstLeaderIP, cluster.getBrokers().get(secondLeader.getNodeId()));
+    setupNetworkPartition(firstLeaderIP, CLUSTER.getBrokers().get(secondLeader.getNodeId()));
 
     // when
     final var future = testCase.when(client);
-    removeNetworkPartition(firstLeaderIP, cluster.getBrokers().get(secondLeader.getNodeId()));
+    clearUnreachableRoutes(CLUSTER.getBrokers().get(secondLeader.getNodeId()));
 
     // then
     testCase.then(client, future);
@@ -132,7 +144,7 @@ final class AsymmetricNetworkPartitionIT {
   }
 
   private void triggerRebalancing() {
-    final var gateway = cluster.getGateways().values().stream().findFirst().orElseThrow();
+    final var gateway = CLUSTER.getGateways().values().stream().findFirst().orElseThrow();
     final var monitoringAddress = gateway.getExternalMonitoringAddress();
     final var httpClient = HttpClient.newHttpClient();
     final var request =
@@ -165,33 +177,13 @@ final class AsymmetricNetworkPartitionIT {
   }
 
   /**
-   * Removes the unreachable `ip route` added via {@link #setupNetworkPartition(String,
-   * ZeebeBrokerNode)}
-   */
-  private void removeNetworkPartition(final String ip, final ZeebeBrokerNode<?> brokerNode)
-      throws IOException, InterruptedException {
-    final var execResult = runCommandInContainer(brokerNode, "ip route del unreachable " + ip);
-    LOGGER.info("{}", execResult.getStdout());
-  }
-
-  /**
    * Set the given ip address unreachable for the given container.
    *
    * @param ipAddress the ip address which should be unreachable
    * @param brokerNode the broker container which should be updated
-   * @throws IOException Can be thrown during running commands in the container
-   * @throws InterruptedException Can be thrown during running commands in the container
    */
-  private void setupNetworkPartition(
-      final String ipAddress, final ZeebeBrokerNode<? extends GenericContainer<?>> brokerNode)
-      throws IOException, InterruptedException {
-    runCommandInContainer(brokerNode, "apt -qq update");
-    runCommandInContainer(brokerNode, "apt install -qq -y iproute2");
-
-    final var execResult =
-        runCommandInContainer(brokerNode, "ip route add unreachable " + ipAddress);
-
-    LOGGER.info("{}", execResult.getStdout());
+  private void setupNetworkPartition(final String ipAddress, final ZeebeNode<?> brokerNode) {
+    exec(brokerNode, "ip route add unreachable " + ipAddress);
   }
 
   private String getContainerNetworkIP(final ZeebeBrokerNode<?> node) {
@@ -201,35 +193,53 @@ final class AsymmetricNetworkPartitionIT {
         .getIpAddress();
   }
 
-  private ExecResult runCommandInContainer(final ZeebeBrokerNode<?> container, final String command)
-      throws IOException, InterruptedException {
-    LOGGER.info("Run command: {}", command);
-
-    final var commands = command.split(" ");
-    final var execResult = container.execInContainer(commands);
-
-    if (execResult.getExitCode() == 0) {
-      LOGGER.info("Command {} was successful.", command);
-    } else {
-      final var errorMessage =
-          String.format(
-              "Command '%s' failed with code: %d stderr: '%s'",
-              command, execResult.getExitCode(), execResult.getStderr());
-      fail(errorMessage);
-    }
-
-    return execResult;
+  /**
+   * Clears all unreachable routes from the given container, i.e. those added via {@link
+   * #setupNetworkPartition(String, ZeebeNode)}.
+   */
+  private static void clearUnreachableRoutes(final ZeebeNode<?> container) {
+    exec(container, "ip route flush type unreachable");
   }
 
-  private void configureBroker(final ZeebeBrokerNode<?> broker) {
+  /**
+   * Installs required network utilities to block certain hosts. Additionally, backs up the routes
+   * table, so we can restore it between tests.
+   */
+  private static void installNetworkUtilities(final ZeebeNode<?> container) {
+    exec(container, "apt-get -qq update");
+    exec(container, "apt-get -qq -y install iproute2");
+  }
+
+  private static void exec(final ZeebeNode<?> container, final String command) {
+    LOGGER.info("Executing command {} on container {}", command, container.getContainerId());
+
+    try {
+      final var result = container.execInContainer(command.split(" "));
+      assertThat(result.getExitCode())
+          .as(
+              "command [%s] failed with status [%d] and output: [%s]",
+              command, result.getExitCode(), result.getStdout())
+          .isEqualTo(0);
+
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+      LangUtil.rethrowUnchecked(e);
+    }
+
+    LOGGER.info("Executed command {} on container {}", command, container.getContainerId());
+  }
+
+  private static void configureBroker(final ZeebeBrokerNode<?> broker) {
     broker
         .self()
-        .withCreateContainerCmdModifier(this::configureNetAdmin)
+        .withCreateContainerCmdModifier(AsymmetricNetworkPartitionIT::configureNetAdmin)
         .withEnv("ZEEBE_BROKER_NETWORK_MAXMESSAGESIZE", "1MB")
         .withEnv("ZEEBE_BROKER_DATA_LOGSEGMENTSIZE", "16MB");
   }
 
-  private void configureNetAdmin(final CreateContainerCmd command) {
+  private static void configureNetAdmin(final CreateContainerCmd command) {
     final var hostConfig = Optional.ofNullable(command.getHostConfig()).orElse(new HostConfig());
     command.withHostConfig(hostConfig.withCapAdd(Capability.NET_ADMIN));
   }

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
@@ -22,8 +22,17 @@ import io.zeebe.containers.ZeebeBrokerNode;
 import io.zeebe.containers.ZeebeContainer;
 import io.zeebe.containers.cluster.ZeebeCluster;
 import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpRequest.BodyPublishers;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.time.Duration;
 import java.util.stream.Stream;
 import org.agrona.CloseHelper;
+import org.agrona.LangUtil;
+import org.awaitility.Awaitility;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Named;
@@ -49,7 +58,7 @@ final class AsymmetricNetworkPartitionIT {
   private ZeebeClient zeebeClient;
 
   @SuppressWarnings("unused")
-  public static Stream<Arguments> provideTestCases() {
+  static Stream<Arguments> provideTestCases() {
     return Stream.of(
         Arguments.arguments(
             Named.named("Deployment distribution", new DeploymentDistributionTestCase())),
@@ -64,18 +73,22 @@ final class AsymmetricNetworkPartitionIT {
   @DisplayName("Withstand Asymmetric Network Partition")
   @ParameterizedTest(name = "{index}: {0}")
   @MethodSource("provideTestCases")
-  public void shouldWithstandAsymmetricNetworkPartition(
+  void shouldWithstandAsymmetricNetworkPartition(
       final AsymmetricNetworkPartitionTestCase asymmetricNetworkPartitionTestCase)
       throws IOException, InterruptedException {
     // given
     setupZeebeCluster();
     zeebeClient = cluster.newClientBuilder().build();
 
+    // the test only works if the leaders of partition 1 and 3 are different nodes
+    Awaitility.await("partition 1 and 3 have a different leader")
+        .atMost(Duration.ofSeconds(30))
+        .during(Duration.ofSeconds(5))
+        .until(this::hasDifferentLeaderForPartitionsOneAndThree);
+
     final var topology = zeebeClient.newTopologyRequest().send().join();
     final var leaderOfPartitionOne = getPartitionLeader(topology, 1);
     final var leaderOfPartitionThree = getPartitionLeader(topology, 3);
-    assertThat(leaderOfPartitionOne.getNodeId()).isNotEqualTo(leaderOfPartitionThree.getNodeId());
-
     final var ipAddress =
         getContainerIpAddress(getContainerForNodeId(leaderOfPartitionOne.getNodeId()));
     asymmetricNetworkPartitionTestCase.given(zeebeClient);
@@ -100,6 +113,39 @@ final class AsymmetricNetworkPartitionIT {
                     .anyMatch(PartitionInfo::isLeader))
         .findFirst()
         .orElseThrow();
+  }
+
+  private void triggerRebalancing() {
+    final var gateway = cluster.getGateways().values().stream().findFirst().orElseThrow();
+    final var monitoringAddress = gateway.getExternalMonitoringAddress();
+    final var httpClient = HttpClient.newHttpClient();
+    final var request =
+        HttpRequest.newBuilder()
+            .POST(BodyPublishers.noBody())
+            .uri(URI.create("http://" + monitoringAddress + "/actuator/rebalance"))
+            .build();
+
+    final HttpResponse<Void> response;
+    try {
+      response = httpClient.send(request, BodyHandlers.discarding());
+      assertThat(response.statusCode()).isEqualTo(200);
+    } catch (final Exception e) {
+      LangUtil.rethrowUnchecked(e);
+    }
+  }
+
+  private boolean hasDifferentLeaderForPartitionsOneAndThree() {
+    final var topology = zeebeClient.newTopologyRequest().send().join();
+    final var firstLeader = getPartitionLeader(topology, 1);
+    final var thirdLeader = getPartitionLeader(topology, 3);
+
+    if (firstLeader.getNodeId() == thirdLeader.getNodeId()) {
+      LOGGER.info("Leader of partition 1 and 3 is {}, re-balancing...", firstLeader.getNodeId());
+      triggerRebalancing();
+      return false;
+    }
+
+    return true;
   }
 
   private void removeAsymmetricNetworkPartition(

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/AsymmetricNetworkPartitionIT.java
@@ -12,6 +12,7 @@ import static org.assertj.core.api.Assertions.fail;
 
 import com.github.dockerjava.api.command.CreateContainerCmd;
 import com.github.dockerjava.api.model.Capability;
+import com.github.dockerjava.api.model.HostConfig;
 import io.camunda.zeebe.client.ZeebeClient;
 import io.camunda.zeebe.client.api.response.BrokerInfo;
 import io.camunda.zeebe.client.api.response.PartitionInfo;
@@ -19,7 +20,6 @@ import io.camunda.zeebe.client.api.response.Topology;
 import io.camunda.zeebe.test.util.testcontainers.ContainerLogsDumper;
 import io.camunda.zeebe.test.util.testcontainers.ZeebeTestContainerDefaults;
 import io.zeebe.containers.ZeebeBrokerNode;
-import io.zeebe.containers.ZeebeContainer;
 import io.zeebe.containers.cluster.ZeebeCluster;
 import java.io.IOException;
 import java.net.URI;
@@ -29,11 +29,13 @@ import java.net.http.HttpRequest.BodyPublishers;
 import java.net.http.HttpResponse;
 import java.net.http.HttpResponse.BodyHandlers;
 import java.time.Duration;
+import java.util.Optional;
 import java.util.stream.Stream;
 import org.agrona.CloseHelper;
 import org.agrona.LangUtil;
 import org.awaitility.Awaitility;
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Named;
 import org.junit.jupiter.api.extension.RegisterExtension;
@@ -44,20 +46,34 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testcontainers.containers.Container.ExecResult;
 import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
 
+@Testcontainers
 final class AsymmetricNetworkPartitionIT {
-
   private static final Logger LOGGER = LoggerFactory.getLogger(AsymmetricNetworkPartitionIT.class);
-  private ZeebeCluster cluster;
+
+  private final Network network = Network.newNetwork();
+
+  @Container
+  private final ZeebeCluster cluster =
+      ZeebeCluster.builder()
+          .withImage(ZeebeTestContainerDefaults.defaultTestImage())
+          .withBrokersCount(3)
+          .withEmbeddedGateway(true)
+          .withPartitionsCount(2)
+          .withReplicationFactor(3)
+          .withBrokerConfig(this::configureBroker)
+          .withNetwork(network)
+          .build();
 
   @SuppressWarnings("unused")
   @RegisterExtension
-  final ContainerLogsDumper logsWatcher =
-      new ContainerLogsDumper(() -> cluster.getBrokers(), LOGGER);
+  final ContainerLogsDumper logsWatcher = new ContainerLogsDumper(cluster::getBrokers, LOGGER);
 
-  private ZeebeClient zeebeClient;
+  private ZeebeClient client;
 
-  @SuppressWarnings("unused")
   static Stream<Arguments> provideTestCases() {
     return Stream.of(
         Arguments.arguments(
@@ -65,43 +81,43 @@ final class AsymmetricNetworkPartitionIT {
         Arguments.arguments(Named.named("Message correlation", new MessageCorrelationTestCase())));
   }
 
+  @BeforeEach
+  void beforeEach() {
+    client = cluster.newClientBuilder().build();
+  }
+
   @AfterEach
   void afterEach() {
-    CloseHelper.quietCloseAll(cluster, zeebeClient);
+    CloseHelper.quietCloseAll(client, network);
   }
 
   @DisplayName("Withstand Asymmetric Network Partition")
   @ParameterizedTest(name = "{index}: {0}")
   @MethodSource("provideTestCases")
-  void shouldWithstandAsymmetricNetworkPartition(
-      final AsymmetricNetworkPartitionTestCase asymmetricNetworkPartitionTestCase)
+  void shouldWithstandAsymmetricNetworkPartition(final AsymmetricNetworkPartitionTestCase testCase)
       throws IOException, InterruptedException {
     // given
-    setupZeebeCluster();
-    zeebeClient = cluster.newClientBuilder().build();
 
     // the test only works if the leaders of partition 1 and 3 are different nodes
-    Awaitility.await("partition 1 and 3 have a different leader")
+    Awaitility.await("partitions have a different leader")
         .atMost(Duration.ofSeconds(30))
         .during(Duration.ofSeconds(5))
-        .until(this::hasDifferentLeaderForPartitionsOneAndThree);
+        .until(this::hasEvenLeaderDistribution);
 
-    final var topology = zeebeClient.newTopologyRequest().send().join();
-    final var leaderOfPartitionOne = getPartitionLeader(topology, 1);
-    final var leaderOfPartitionThree = getPartitionLeader(topology, 3);
-    final var ipAddress =
-        getContainerIpAddress(getContainerForNodeId(leaderOfPartitionOne.getNodeId()));
-    asymmetricNetworkPartitionTestCase.given(zeebeClient);
-    setupAsymmetricNetworkPartition(
-        ipAddress, getContainerForNodeId(leaderOfPartitionThree.getNodeId()));
+    final var topology = client.newTopologyRequest().send().join();
+    final var firstLeader = getPartitionLeader(topology, 1);
+    final var secondLeader = getPartitionLeader(topology, 2);
+    final var firstLeaderIP =
+        getContainerNetworkIP(cluster.getBrokers().get(firstLeader.getNodeId()));
+    testCase.given(client);
+    setupNetworkPartition(firstLeaderIP, cluster.getBrokers().get(secondLeader.getNodeId()));
 
     // when
-    final var future = asymmetricNetworkPartitionTestCase.when(zeebeClient);
-    removeAsymmetricNetworkPartition(
-        ipAddress, getContainerForNodeId(leaderOfPartitionThree.getNodeId()));
+    final var future = testCase.when(client);
+    removeNetworkPartition(firstLeaderIP, cluster.getBrokers().get(secondLeader.getNodeId()));
 
     // then
-    asymmetricNetworkPartitionTestCase.then(zeebeClient, future);
+    testCase.then(client, future);
   }
 
   private BrokerInfo getPartitionLeader(final Topology topology, final int partition) {
@@ -134,13 +150,13 @@ final class AsymmetricNetworkPartitionIT {
     }
   }
 
-  private boolean hasDifferentLeaderForPartitionsOneAndThree() {
-    final var topology = zeebeClient.newTopologyRequest().send().join();
+  private boolean hasEvenLeaderDistribution() {
+    final var topology = client.newTopologyRequest().send().join();
     final var firstLeader = getPartitionLeader(topology, 1);
-    final var thirdLeader = getPartitionLeader(topology, 3);
+    final var otherLeader = getPartitionLeader(topology, 2);
 
-    if (firstLeader.getNodeId() == thirdLeader.getNodeId()) {
-      LOGGER.info("Leader of partition 1 and 3 is {}, re-balancing...", firstLeader.getNodeId());
+    if (firstLeader.getNodeId() == otherLeader.getNodeId()) {
+      LOGGER.info("Leader of all partitions is {}, re-balancing...", firstLeader.getNodeId());
       triggerRebalancing();
       return false;
     }
@@ -148,12 +164,14 @@ final class AsymmetricNetworkPartitionIT {
     return true;
   }
 
-  private void removeAsymmetricNetworkPartition(
-      final String ipAddress, final ZeebeBrokerNode<? extends GenericContainer<?>> brokerNode)
+  /**
+   * Removes the unreachable `ip route` added via {@link #setupNetworkPartition(String,
+   * ZeebeBrokerNode)}
+   */
+  private void removeNetworkPartition(final String ip, final ZeebeBrokerNode<?> brokerNode)
       throws IOException, InterruptedException {
-    LOGGER.info(
-        "{}",
-        runCommandInContainer(brokerNode, "ip route del unreachable " + ipAddress).getStdout());
+    final var execResult = runCommandInContainer(brokerNode, "ip route del unreachable " + ip);
+    LOGGER.info("{}", execResult.getStdout());
   }
 
   /**
@@ -164,35 +182,26 @@ final class AsymmetricNetworkPartitionIT {
    * @throws IOException Can be thrown during running commands in the container
    * @throws InterruptedException Can be thrown during running commands in the container
    */
-  private void setupAsymmetricNetworkPartition(
+  private void setupNetworkPartition(
       final String ipAddress, final ZeebeBrokerNode<? extends GenericContainer<?>> brokerNode)
       throws IOException, InterruptedException {
-    runCommandInContainer(brokerNode, "apt update");
-    runCommandInContainer(brokerNode, "apt install -y iproute2");
-    LOGGER.info(
-        "{}",
-        runCommandInContainer(brokerNode, "ip route add unreachable " + ipAddress).getStdout());
+    runCommandInContainer(brokerNode, "apt -qq update");
+    runCommandInContainer(brokerNode, "apt install -qq -y iproute2");
+
+    final var execResult =
+        runCommandInContainer(brokerNode, "ip route add unreachable " + ipAddress);
+
+    LOGGER.info("{}", execResult.getStdout());
   }
 
-  private ZeebeBrokerNode<? extends GenericContainer<?>> getContainerForNodeId(final int nodeId) {
-    return cluster.getBrokers().get(nodeId);
-  }
-
-  private String getContainerIpAddress(
-      final ZeebeBrokerNode<? extends GenericContainer<?>> leaderOneNode) {
-    return leaderOneNode
-        .getCurrentContainerInfo()
-        .getNetworkSettings()
-        .getNetworks()
-        .values()
-        .stream()
+  private String getContainerNetworkIP(final ZeebeBrokerNode<?> node) {
+    return node.getContainerInfo().getNetworkSettings().getNetworks().values().stream()
         .findFirst()
         .orElseThrow()
         .getIpAddress();
   }
 
-  private ExecResult runCommandInContainer(
-      final ZeebeBrokerNode<? extends GenericContainer<?>> container, final String command)
+  private ExecResult runCommandInContainer(final ZeebeBrokerNode<?> container, final String command)
       throws IOException, InterruptedException {
     LOGGER.info("Run command: {}", command);
 
@@ -212,30 +221,16 @@ final class AsymmetricNetworkPartitionIT {
     return execResult;
   }
 
-  /** Set ups Zeebe Cluster with necessary capabilities in order to create network partitions. */
-  private void setupZeebeCluster() {
-    cluster =
-        ZeebeCluster.builder()
-            .withImage(ZeebeTestContainerDefaults.defaultTestImage())
-            .withBrokersCount(3)
-            .withEmbeddedGateway(true)
-            .withPartitionsCount(3)
-            .withReplicationFactor(3)
-            .build();
-    cluster
-        .getBrokers()
-        .forEach(
-            (nodeId, broker) -> {
-              ((ZeebeContainer) broker)
-                  .withCreateContainerCmdModifier(
-                      (CreateContainerCmd it) ->
-                          it.withHostConfig(it.getHostConfig().withCapAdd(Capability.NET_ADMIN)));
+  private void configureBroker(final ZeebeBrokerNode<?> broker) {
+    broker
+        .self()
+        .withCreateContainerCmdModifier(this::configureNetAdmin)
+        .withEnv("ZEEBE_BROKER_NETWORK_MAXMESSAGESIZE", "1MB")
+        .withEnv("ZEEBE_BROKER_DATA_LOGSEGMENTSIZE", "16MB");
+  }
 
-              // smaller sizes to make the tests faster
-              broker
-                  .withEnv("ZEEBE_BROKER_NETWORK_MAXMESSAGESIZE", "1MB")
-                  .withEnv("ZEEBE_BROKER_DATA_LOGSEGMENTSIZE", "16MB");
-            });
-    cluster.start();
+  private void configureNetAdmin(final CreateContainerCmd command) {
+    final var hostConfig = Optional.ofNullable(command.getHostConfig()).orElse(new HostConfig());
+    command.withHostConfig(hostConfig.withCapAdd(Capability.NET_ADMIN));
   }
 }

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/MessageCorrelationTestCase.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/network/MessageCorrelationTestCase.java
@@ -30,9 +30,9 @@ final class MessageCorrelationTestCase implements AsymmetricNetworkPartitionTest
             .done();
     client.newDeployResourceCommand().addProcessModel(process, "process.bpmn").send().join();
     // make sure that the message is correlated to correct partition
-    assertThat(SubscriptionUtil.getSubscriptionPartitionId(BufferUtil.wrapString("key"), 3))
-        .describedAs("message should correlated to partition three")
-        .isEqualTo(3);
+    assertThat(SubscriptionUtil.getSubscriptionPartitionId(BufferUtil.wrapString("key"), 2))
+        .describedAs("message should correlated to partition two")
+        .isEqualTo(2);
 
     client
         .newPublishMessageCommand()

